libraries/chain/webassembly/eos-vm-oc/executor.cpp

```diff
@@ -0,0 +1,251 @@
+#include <eosio/chain/webassembly/eos-vm-oc/executor.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/code_cache.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/memory.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/intrinsic_mapping.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/intrinsic.hpp>
+#include <eosio/chain/wasm_eosio_constraints.hpp>
+#include <eosio/chain/apply_context.hpp>
+#include <eosio/chain/transaction_context.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/types.hpp>
+
+#include <fc/scoped_exit.hpp>
+
+#include <boost/hana/equal.hpp>
+
+#include <mutex>
+
+#include <asm/prctl.h>
+#include <sys/prctl.h>
+#include <sys/syscall.h>
+
+extern "C" int arch_prctl(int code, unsigned long* addr);
+
+namespace eosio { namespace chain { namespace eosvmoc {
+
+static constexpr auto signal_sentinel = 0x4D56534F45534559ul;
+
+static std::mutex inited_signal_mutex;
+static bool inited_signal;
+
+static void(*chained_handler)(int,siginfo_t*,void*);
+[[noreturn]] static void segv_handler(int sig, siginfo_t* info, void* ctx)  {
+   control_block* cb_in_main_segment;
+
+   //a 0 GS value is an indicator an executor hasn't been active on this thread recently
+   uint64_t current_gs;
+   syscall(SYS_arch_prctl, ARCH_GET_GS, &current_gs);
+   if(current_gs == 0)
+      goto notus;
+
+   cb_in_main_segment = reinterpret_cast<control_block*>(current_gs - memory::cb_offset);
+
+   //as a double check that the control block pointer is what we expect, look for the magic
+   if(cb_in_main_segment->magic != signal_sentinel)
+      goto notus;
```

Comment:
If this were to trigger, there's no guarantee that cb_in_main_segment points to valid memory.

Reply:
Likewise for anything else that accesses cb_in_main_segment in here. Are you suggesting that it's a worthless check and might as well be removed?

Reply:
I presumed that the intent of this check is to determine whether gs is ours.  Thus, the later checks should be safe provided this passes.  The check here is not entirely worthless, but it isn't very reliable either.